lib/services/fanout.go

```diff
@@ -325,16 +342,38 @@ type fanoutWatcher struct {
 }
 
 // init transmits the OpInit event.  safe to double-call.
-func (w *fanoutWatcher) init() (ok bool) {
+// Validates requested resource kinds against the list confirmed by the upstream event source.
+// See Fanout.SetInit() for more details.
+func (w *fanoutWatcher) init(confirmedKinds map[kindSubKind]types.WatchKind) (err error) {
 	w.initOnce.Do(func() {
+		w.initOk = false
+
+		var validKinds []types.WatchKind
```

Comment:
nit: the fanout layer often falls under tremendous load. This is somewhere that pre-allocating slices with the correct capacity makes a real perf difference. Ex:

```golang
validKinds := make([]types.WatchKind, 0, len(w.watch.Kinds))
```

Reply:
Makes sense. I added pre-allocation to this and 2 more places in https://github.com/gravitational/teleport/pull/24080/commits/471f539ac3796446776a0079ddd59e33268203ef

If it really makes a difference, I can optimize this further for the best case  (and the most common) scenario when `validKinds == w.watch.Kinds` at the end. We'd just track rejected kinds using a map that would only be allocated on the first write. At the end, if it's empty we'd just copy `w.watch.Kinds` to `validKinds` with 0 allocations. In any other than the best case it'd be 2 allocations though.